src/Easify.NoSql.MongoDb/MongoDbRepository.cs

```diff
@@ -0,0 +1,307 @@
+// 
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+// 
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// 
+
+
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Easify.Extensions.Specifications;
+using Easify.NoSql.ComponentModel;
+using MongoDB.Driver;
+
+namespace Easify.NoSql.MongoDb
+{
+    public class MongoDbRepository<T> : IMongoDbRepository<T>
+    {
+        public MongoDbRepository(IMongoDatabase mongoDatabase, string mongoCollectionName)
+        {
+            var name = mongoCollectionName ?? throw new ArgumentNullException(nameof(mongoCollectionName));
+            MongoDatabase = mongoDatabase ?? throw new ArgumentNullException(nameof(mongoDatabase));
+            MongoCollection = mongoDatabase.GetCollection<T>(name);
+        }
+
+        protected IMongoCollection<T> MongoCollection { get; }
+
+        protected IMongoDatabase MongoDatabase { get; }
+
+        public virtual Task InsertAsync(T t)
+        {
+            return MongoCollection.InsertOneAsync(t);
+        }
+
+        public virtual Task InsertManyAsync(T[] ts)
+        {
+            return MongoCollection.InsertManyAsync(ts);
+        }
+
+        public virtual Task<T> GetAsync(Expression<Func<T, bool>> condition)
+        {
+            return InternalGetAsync(condition, m => m, CancellationToken.None);
+        }
+
+        public virtual Task<T> GetAsync(Func<IFindFluent<T, T>, IFindFluent<T, T>> query)
+        {
+            return InternalGetAsync(m => true, query, CancellationToken.None);
+        }
+
+        public virtual Task<T> GetAsync(Specification<T> specification,
+            Func<IFindFluent<T, T>, IFindFluent<T, T>> query)
+        {
+            return InternalGetAsync(specification?.ToExpression(), query, CancellationToken.None);
+        }
+
+        public virtual Task<T> GetAsync(Specification<T> specification)
+        {
+            return InternalGetAsync(specification?.ToExpression(), m => m, CancellationToken.None);
+        }
+
+        public virtual Task<T> GetAsync(Expression<Func<T, bool>> condition,
+            Func<IFindFluent<T, T>, IFindFluent<T, T>> query)
+        {
+            return InternalGetAsync(condition, query, CancellationToken.None);
+        }
+
+        public virtual Task<T> GetAsync(Expression<Func<T, bool>> condition,
+            Func<IFindFluent<T, T>, IFindFluent<T, T>> query,
+            CancellationToken cancellationToken)
+        {
+            return InternalGetAsync(condition, query, cancellationToken);
+        }
+
+        [Obsolete("Use function with IFindFluent interface")]
```

Comment:
As discussed, it'd be good not to have to bring Obsolete methods across

Reply:
I made some refactoring here but removed the Obsolete annotation as they were necessary